ktor-io/common/src/io/ktor/utils/io/ByteChannel.kt

```diff
@@ -133,65 +135,99 @@ public class ByteChannel(public val autoFlush: Boolean = false) : ByteReadChanne
+        expectedSlot.resume()
     }
 
     /**
      * Cancel waiter.
      */
     private fun closeSlot(cause: Throwable?) {
-        val closeContinuation = if (cause != null) ClosedSlot(cause) else io.ktor.utils.io.internal.CLOSED
-        val continuation = suspensionSlot.getAndSet(closeContinuation) ?: return
-        if (continuation is ClosedSlot) return
+        val closeContinuation = if (cause != null) Slot.Closed(cause) else Slot.CLOSED
+        val continuation = suspensionSlot.getAndSet(closeContinuation)
+        if (continuation is Slot.Closed) return
 
         if (cause != null) {
             continuation.resumeWithException(cause)
         } else {
-            continuation.resume(Unit)
+            continuation.resume()
+        }
+    }
+
+    private inline fun <reified TaskType : Slot> trySuspendSlot(
+        slot: TaskType,
+        crossinline sleepCondition: () -> Boolean
+    ) {
+        // When the slot is full, try resuming the current task
+        if (!suspensionSlot.compareAndSet(Slot.Empty, slot)) {
+            val currentSlot = suspensionSlot.value as? Slot.TaskSlot ?: run {
+                slot.resume()
+                return
+            }
+            if (suspensionSlot.compareAndSet(currentSlot, Slot.Empty)) {
+                when (currentSlot) {
+                    // If the current task is the same, cancel it
+                    is TaskType -> currentSlot.resumeWithException(
+                        IllegalStateException("Concurrent ${slot::class.simpleName?.lowercase()} attempts")
+                    )
+                    else -> currentSlot.resume()
+                }
+            }
+            slot.resume()
+        } else if (!sleepCondition()) {
+            suspensionSlot.getAndSet(Slot.Empty).resume()
+        } else {
+            // suspend the coroutine until another thread resumes
         }
     }
 
-    private inline fun trySuspendSlot(continuation: Continuation<Unit>, crossinline sleepCondition: () -> Boolean) {
-        val published = suspensionSlot.compareAndSet(null, continuation)
-        if (!published) {
-            continuation.resume(Unit)
-            return
+    private sealed class Slot {
+        companion object {
+            val CLOSED = Closed(null)
         }
 
-        if (!sleepCondition()) {
-            suspensionSlot.getAndSet(null)?.resume(Unit)
+        open fun resume() {}
+        open fun resumeWithException(throwable: Throwable) {}
+
+        data object Empty : Slot()
+        data class Closed(val cause: Throwable?) : Slot()
+        abstract class TaskSlot : Slot() {
+            abstract val continuation: Continuation<Unit>
+            override fun resume() =
+                continuation.resume(Unit)
+            override fun resumeWithException(throwable: Throwable) =
+                continuation.resumeWithException(throwable)
         }
+        data class Read(override val continuation: Continuation<Unit>) : TaskSlot()
```

Comment:
Could you tell me why it should be a data class and if we need inheritance?

Reply:
Oh, there's no reason.  I originally had duplication then refactored and didn't drop the data.  I'll drop the data.